File: benchmark/analysis/fit.py
"""Model-fit estimator (#20).

Estimates whether a model variant fits in available memory using model
metadata (parameter count, quantization bits) and standard weight-bytes
arithmetic. Output is clearly labeled as an ESTIMATE with recorded
assumptions — never presented as a measured fact. Measured peak memory
from a real run, when present, always takes precedence.
"""

from __future__ import annotations

import re
from typing import Any

__all__ = ["estimate_model_fit", "parse_parameter_count", "BITS_PER_WEIGHT"]

# Bits per weight by quantization label. These are format conventions,
# not measurements; unknown labels are rejected rather than guessed.
BITS_PER_WEIGHT: dict[str, float] = {
    "fp32": 32.0,
    "fp16": 16.0,
    "bf16": 16.0,
    "f16": 16.0,
    "q8_0": 8.5,
    "q8": 8.5,
    "int8": 8.0,
    "q6_k": 6.5625,
    "q5_k_m": 5.67,
    "q5_k_s": 5.54,
    "q5_0": 5.54,
    "q5_1": 6.7,
    "q4_k_m": 4.85,
    "q4_k_s": 4.58,
    "q4_0": 4.55,
    "q4_1": 5.0,
    "q3_k_m": 3.91,
    "q3_k_s": 3.5,
    "q2_k": 3.35,
    "iq4_xs": 4.25,
    "iq3_xxs": 3.06,
    "iq2_xxs": 2.06,
}

_PARAM_RE = re.compile(r"([\d.]+)\s*([bmk])\b", re.IGNORECASE)
_MULTIPLIERS = {"k": 1e3, "m": 1e6, "b": 1e9}


def parse_parameter_count(text: str | None) -> float | None:
    """Parse '7B', '1.5b', '350M' style parameter counts."""
    if not text:
        return None
    match = _PARAM_RE.search(text)
    if not match:
        return None
    value = float(match.group(1))
    unit = match.group(2).lower()
    return value * _MULTIPLIERS[unit]


def estimate_model_fit(
    parameters_text: str | None,
    quantization: str | None,
    available_vram_mb: float | None = None,
    available_ram_mb: float | None = None,
    context_tokens: int = 4096,
    overhead_factor: float = 1.15,
) -> dict[str, Any]:
    """Estimate memory need vs availability.

    ``overhead_factor`` covers KV cache, activations, and runtime overhead
    beyond raw weights; it is an assumption, recorded in the output.
    """
    params = parse_parameter_count(parameters_text)
    if params is None:
        return {
            "estimated_weights_gb": None,
            "estimated_total_gb": None,
            "fits": None,
            "reason": "parameter count unavailable; cannot estimate",
            "assumptions": {},
        }
    bits = BITS_PER_WEIGHT.get((quantization or "").lower())
    if bits is None:
        return {
            "estimated_weights_gb": None,
            "estimated_total_gb": None,
            "fits": None,
            "reason": f"unknown quantization {quantization!r}; refusing to guess bits-per-weight",
            "assumptions": {"known_quantizations": sorted(BITS_PER_WEIGHT)},
        }
    weights_bytes = params * bits / 8.0
    total_bytes = weights_bytes * overhead_factor
    total_gb = total_bytes / 1e9
    assumptions = {
        "bits_per_weight": bits,
        "overhead_factor": overhead_factor,
        "context_tokens": context_tokens,
        "note": (
            "estimate from parameter count x bits/weight x overhead factor; "
            "KV cache scales with context and is approximated by the factor"
        ),
    }
    candidates: list[tuple[str, float]] = []
    if available_vram_mb is not None:
        candidates.append(("vram", available_vram_mb / 1000.0))
    if available_ram_mb is not None:
        candidates.append(("ram", available_ram_mb / 1000.0))
    fits: bool | None = None
    target: str | None = None
    for name, gb in candidates:
        if gb >= total_gb:
            fits = True
            target = name
            break
        fits = False
        target = name
    return {
        "estimated_weights_gb": round(weights_bytes / 1e9, 3),
        "estimated_total_gb": round(total_gb, 3),
        "fits": fits,
        "fit_target": target,
        "reason": (
            f"estimated {total_gb:.2f} GB needed vs "
            + ", ".join(f"{n} {gb:.1f} GB" for n, gb in candidates)
            if candidates
            else "no memory availability supplied"
        ),
        "assumptions": assumptions,
    }

File: web/src/lib/fit.ts
// Browser-side model-fit estimator.
//
// This mirrors `aihwbench/analysis/fit.py`, which stays canonical. The
// bits-per-weight table and the overhead factor are not duplicated here: they
// are generated into `data/constants.json` from the Python module, and that
// file also carries reference vectors the tests check this implementation
// against, so the two cannot drift apart silently.
//
// Everything it produces is an ESTIMATE. A measured result for the same model
// on the same hardware always beats it, and the UI says so.

import type { FitConstants } from './types'

export interface FitEstimate {
  estimated_weights_gb: number | null
  estimated_total_gb: number | null
  fits: boolean | null
  fit_target: string | null
  reason: string
}

const PARAM_RE = /([\d.]+)\s*([bmk])\b/i
const MULTIPLIERS: Record<string, number> = { k: 1e3, m: 1e6, b: 1e9 }

/** Parse '7B', '1.5b', '350M' style parameter counts. */
export function parseParameterCount(text: string | null | undefined): number | null {
  if (!text) return null
  const match = PARAM_RE.exec(text)
  const value = match?.[1]
  const unit = match?.[2]
  if (value === undefined || unit === undefined) return null
  const multiplier = MULTIPLIERS[unit.toLowerCase()]
  if (multiplier === undefined) return null
  return parseFloat(value) * multiplier
}

function round(value: number, places: number): number {
  const factor = 10 ** places
  return Math.round(value * factor) / factor
}

/**
 * Estimate memory need against what the machine has.
 *
 * An unknown quantization returns `fits: null` with a reason rather than a
 * guess: inventing a bits-per-weight figure would produce a confident answer
 * from nothing, which is the failure this project exists to avoid.
 */
export function estimateModelFit(
  constants: FitConstants,
  parametersText: string | null | undefined,
  quantization: string | null | undefined,
  availableVramMb: number | null = null,
  availableRamMb: number | null = null,
): FitEstimate {
  const params = parseParameterCount(parametersText)
  if (params === null) {
    return {
      estimated_weights_gb: null,
      estimated_total_gb: null,
      fits: null,
      fit_target: null,
      reason: 'parameter count unavailable; cannot estimate',
    }
  }

  const bits = constants.bits_per_weight[(quantization ?? '').toLowerCase()]
  if (bits === undefined) {
    return {
      estimated_weights_gb: null,
      estimated_total_gb: null,
      fits: null,
      fit_target: null,
      reason: `unknown quantization '${quantization}'; refusing to guess bits-per-weight`,
    }
  }

  const weightsBytes = (params * bits) / 8.0
  const totalGb = (weightsBytes * constants.overhead_factor) / 1e9

  const candidates: [string, number][] = []
  if (availableVramMb !== null) candidates.push(['vram', availableVramMb / 1000.0])
  if (availableRamMb !== null) candidates.push(['ram', availableRamMb / 1000.0])

  let fits: boolean | null = null
  let target: string | null = null
  for (const [name, gb] of candidates) {
    if (gb >= totalGb) {
      fits = true
      target = name
      break
    }
    fits = false
    target = name
  }

  return {
    estimated_weights_gb: round(weightsBytes / 1e9, 3),
    estimated_total_gb: round(totalGb, 3),
    fits,
    fit_target: target,
    reason:
      candidates.length > 0
        ? `estimated ${totalGb.toFixed(2)} GB needed vs ` +
          candidates.map(([n, gb]) => `${n} ${gb.toFixed(1)} GB`).join(', ')
        : 'no memory availability supplied',
  }
}

/**
 * How much of the model has to live outside VRAM.
 *
 * The single biggest performance cliff in local inference is the moment a
 * model spills from GPU to system RAM, and the size of the spill is what
 * decides how far throughput falls. This reports the ratio; it deliberately
 * does not predict a speed, because that depends on the machine and is what
 * the measured dataset is for.
 */
export function offloadFraction(
  totalGb: number | null,
  availableVramMb: number | null,
): number | null {
  if (totalGb === null || availableVramMb === null || totalGb <= 0) return null
  const vramGb = availableVramMb / 1000.0
  if (vramGb >= totalGb) return 0
  return round((totalGb - vramGb) / totalGb, 3)
}
